feat(ivy): convert [ngStyle] and [ngClass] to use ivy styling bindings

matsko opened the pull request:
Prior to this fix, both the `NgStyle` and `NgClass` directives made use
of `Renderer2` and this dependency raised issues for future versions of
Angular that cannot inject it. This patch ensures that there are two
versions of both directives: one for the VE and another for Ivy.

matsko committed:
feat(ivy): open up ivy_switch_mode to non-core packages

Prior to this fix, using the compiler's ivy_switch mechanism was
only available to core packages. This patch allows for this variable
switching mechanism to work across all other angular packages.

matsko commented:
https://tap.corp.google.com/ui#id=OCL:234186770:BASE:234205915:1550267312708:c5096264

gkalpak commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -202,8 +202,7 @@ export function extractDirectiveMetadata(
   // Detect if the component inherits from another class
-  const usesInheritance = clazz.heritageClauses !== undefined &&
-      clazz.heritageClauses.some(hc => hc.token === ts.SyntaxKind.ExtendsKeyword);
+  const usesInheritance = reflector.hasBaseClass(clazz);
```
This has now been fixed in master (via #28773).

matsko committed:
feat(ivy): convert [ngStyle] and [ngClass] to use ivy styling bindings

Prior to this fix, both the `NgStyle` and `NgClass` directives made use
of `Renderer2` and this dependency raised issues for future versions of
Angular that cannot inject it. This patch ensures that there are two
versions of both directives: one for the VE and another for Ivy.

Jira Issue: FW-882

mhevery commented on packages/common/src/directives/ng_class_impl.ts:
```diff
@@ -0,0 +1,208 @@
+/**
```
Any particular reason why this code was pulled into a separate file?

mhevery commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+ * Used to diff and convert ngStyle/ngClass instructions into [style] and [class] bindings.
+ *
+ * ngStyle and ngClass both accept various forms of input and behave differently than that
+ * of how [style] and [class] behave in Angluar.
```
```suggestion
 * of how [style] and [class] behave in Angular.
```

mhevery commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+ * into something that the core-level [style] and [class] bindings understand.
+ *
+ * This [StylingDiffer] class handles this conversion by creating a new input value each time
+ * the inner guts of the binding value have changed.
```
```suggestion
 * the inner representation of the binding value have changed.
```

mhevery commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+ * respect and understand the styling values emitted through ngStyle and ngClass (when they
+ * are present and used in a template).
+ *
+ * Instead of giving these directives a pass, they should be included into the styling algorithm
```
what do you mean `giving a pass`? could you rewrite in more clear way?

mhevery commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+ *
+ * Here's why:
+ *
+ * - if ngStyle/ngClass is used in combination with [style]/[class] bindings then strange things
```
what strange things?

mhevery commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+ *
+ * - if ngStyle/ngClass is used in combination with [style]/[class] bindings then strange things
+ *   will occur in regards to prioritization.
+ * - ngClass/ngStyle do not respect [class.name] and [style.prop] bindings (they will write
```
give an example. Right now the reader has no way of guessing wha the `given the right combination of events` are

mhevery commented on packages/core/src/render3/index.ts:
```diff
@@ -19,6 +19,7 @@ export {RenderFlags} from './interfaces/definition';
 export {CssSelectorList} from './interfaces/projection';
 
 
+
```
revert

mhevery commented on tools/public_api_guard/common/common.d.ts:
```diff
@@ -207,12 +207,12 @@ export declare class LowerCasePipe implements PipeTransform {
 }
 
-export declare class NgClass implements DoCheck {
+export declare class NgClass extends NgClassBase {
```
Something is wrong here. The public API reffers to `NgClassBase` but it is not shown here. Same issue for  `NgStyleBase`

mhevery commented on packages/common/src/directives/ng_style.ts:
```diff
@@ -5,8 +5,65 @@
+ * If the VE is present (which is for older versions of Angular) then NgStyle will inject
+ * the legacy diffing algorithm as a service and delegate all styling changes to that.
+ *
+ * If the VE is not present then NgClass will normailze (through the injected service) and
```
```suggestion
 * If the VE is not present then NgClass will normalize (through the injected service) and
```

mhevery commented on packages/common/src/directives/ng_style.ts:
```diff
@@ -5,8 +5,65 @@
+ * the legacy diffing algorithm as a service and delegate all styling changes to that.
+ *
+ * If the VE is not present then NgClass will normailze (through the injected service) and
+ * then write all styling changes to the `[class]` binding directly (through a host binding).
```
```suggestion
 * then write all styling changes to the `[style]` binding directly (through a host binding).
```

matsko committed:
fixup! feat(ivy): convert [ngStyle] and [ngClass] to use ivy styling bindings
